### PR TITLE
Symbol manglers refactor

### PR DESCRIPTION
**Pull request by kkosowski-bt**

### CODE REVIEW

File: MachObfuscator.xcodeproj/xcshareddata/xcschemes/MachObfuscator.xcscheme [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+         <CommandLineArgument
+            argument = "-v"
+            isEnabled = "NO">
+         </CommandLineArgument>
+         <CommandLineArgument
+            argument = "-m realWords"
+            isEnabled = "YES">
+         </CommandLineArgument>
+         <CommandLineArgument
+            argument = "/Users/kamilkosowski/Documents/OSXDemo-kopia/OSXDemo.app"
```
Comment (kam800) [inline MachObfuscator.xcodeproj/xcshareddata/xcschemes/MachObfuscator.xcscheme]: Your local path here. Do you think we need a shared schema at all? Each one of us would use different parameters anyway.

File: MachObfuscator/main.swift [hunk 2]
```diff
@@ -5,7 +5,6 @@ private func main() {
     guard !options.help,
         let appDirectory = options.appDirectory,
         let mangler = SymbolManglers.mangler(byKey: options.manglerKey) else {
-        print(Options.usage)
```
Comment (kam800) [inline MachObfuscator/main.swift]: Wyh did you remove that? I always thought printing usage description on incorrect arguments is a standard.

File: MachObfuscator/SymbolMangling/RealWordsMangler/RealWordsMangler.swift [hunk 3]
```diff
@@ -1,12 +1,15 @@
 import Foundation
 
-class RealWordsMangler: SymbolMangling {
-    static var key: String = "realWords"
-    static let helpDescription: String = "replace objc symbols with random words (dyld info obfuscation not supported yet)"
+final class RealWordsMangler: SymbolMangling {
+    var key: String = "realWords"
 
-    private let exportTrieMangler: ExportTrieMangling = ExportTrieMangler()
+    let helpDescription: String = "replace objc symbols with random words (dyld info obfuscation not supported yet)"
```
Comment (kam800) [inline MachObfuscator/SymbolMangling/RealWordsMangler/RealWordsMangler.swift]:
```suggestion
    let helpDescription: String = "replace objc symbols with random words and fill dyld info symbols with numbers"
```

File: MachObfuscator/SymbolMangling/CezarMangler/CaesarCypher.swift [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import Foundation
+
+final class CaesarCypher {
+    private let asciiRange = UInt8(33) ... UInt8(126)
+
+    func cypher(element: UInt8, cypherKey: UInt8) -> UInt8 {
```
Comment (kam800) [inline MachObfuscator/SymbolMangling/CezarMangler/CaesarCypher.swift]: What do you think about using some verb?
```suggestion
    func encrypt(element: UInt8, key: UInt8) -> UInt8 {
```

File: MachObfuscator/SymbolMangling/CezarMangler/CaesarMangler.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import Foundation
+
+final class CaesarMangler: SymbolMangling {
+    // todo: Probably should be changed to 'caesar'. Would require readme.md update
```
Comment (kam800) [inline MachObfuscator/SymbolMangling/CezarMangler/CaesarMangler.swift]: I agree, that typo puts a shame on me :) What about updating this key and readme.md in this PR?

File: MachObfuscator/Options/Options.swift [hunk 6]
```diff
@@ -17,12 +17,12 @@ extension Options {
                          argv: CommandLine.arguments)
     }
 
-    init(argc: Int32, unsafeArgv: UnsafeArgv, argv: [String]) {
+    init(argc: Int32, unsafeArgv: UnsafeArgv, argv: [String], defaultManglerKey: String = "realWords") {
```
Comment (kam800) [inline MachObfuscator/Options/Options.swift]: What about moving `defaultManglerKey` to `static let defaultManglerKey = RealWordsMangler.key`? It is a little unsafe to use string literal when the const is available.

File: MachObfuscator/SymbolMangling/SymbolManglingMap.swift [hunk 7]
```diff
@@ -1,7 +1,13 @@
 import Foundation
 
 struct SymbolManglingMap {
+    typealias ObfuscationTriePair = (unobfuscated: Trie, obfuscated: Trie)
+
+    typealias TriesPerCpu = [CpuId: ObfuscationTriePair]
```
[the remark this answers is not in the document]
Comment (kkosowski-bt) [reply MachObfuscator/SymbolMangling/SymbolManglingMap.swift]: My motivation for plural was to make it clear, that you have in fact two export tries: unobfuscated and obfuscated, hence the reason I created the typealias `ObfuscationTriePair`. But I also see your point: `TriesPerCpu` could indicate, that there is more than one trie per cpu. and it says nothing about the division: obfuscated, unobfuscated, but `ObfuscationTriePair` makes it clear. Anyway, I believe your comment is valid, the naming could be somehow confusing. I'll change it to singular.

File: MachObfuscator/SymbolMangling/SymbolManglingMap.swift [hunk 8]
```diff
@@ -1,7 +1,13 @@
...
+
     var selectors: [String: String]
+
     var classNames: [String: String]
-    var unobfuscatedObfuscatedTriePairPerCpuIdPerURL: [URL: [CpuId: (Trie, Trie)]]
+
+    var exportTrieObfuscationMap: [URL: TriesPerCpu]
```
[the remark this answers is not in the document]
Comment (kkosowski-bt) [reply MachObfuscator/SymbolMangling/SymbolManglingMap.swift]: You're most welcome, glad you like it :)

File: MachObfuscator/SymbolMangling/CezarMangler/CaesarCypher.swift (hunk 4 above)
Comment (kkosowski-bt) [reply MachObfuscator/SymbolMangling/CezarMangler/CaesarCypher.swift]: hmm there is such verb as `to cypher`. actually, I like both encrypt and cypher, maybe encrypt would be even better visually..? let me know what you think

File: MachObfuscator/SymbolMangling/CezarMangler/CaesarMangler.swift (hunk 5 above)
Comment (kkosowski-bt) [reply MachObfuscator/SymbolMangling/CezarMangler/CaesarMangler.swift]: It's just a typo, no worries! will add it to readme :)

File: MachObfuscatorTests/SymbolManglingTests/ExportTrieMangler_Tests.swift [hunk 9]
```diff
@@ -2,12 +2,12 @@ import XCTest
 
 class ExportTrieMangler_Tests: XCTestCase {
```
[the remark this answers is not in the document]
Comment (kkosowski-bt) [reply MachObfuscatorTests/SymbolManglingTests/ExportTrieMangler_Tests.swift]: Sure, sorry - forgot about it.

File: MachObfuscator/main.swift (hunk 2 above)
Comment (kkosowski-bt) [reply MachObfuscator/main.swift]: Sorry - that's my mistake. Was in a hurry, thought it's my helper print.

File: MachObfuscator/SymbolMangling/SymbolManglers.swift [hunk 10]
```diff
@@ -1,11 +1,13 @@
 import Foundation
 
 enum SymbolManglers {
-    static let allManglers: [SymbolMangling.Type] = [CezarMangler.self, RealWordsMangler.self]
+    static let allManglers: [SymbolMangling] = [CaesarMangler(exportTrieMangler: CaesarExportTrieMangler()),
+                                                RealWordsMangler(exportTrieMangler: RealWordsExportTrieMangler())]
+
+    static let defaultManglerKey = "realWords"
```
Comment (kam800) [inline MachObfuscator/SymbolMangling/SymbolManglers.swift]: I am glad you moved this const out of `Options`, but I am still worried about using `"realWords"` string literal here, while it is used in `RealWordsMangler` as:
```
   var key: String = "realWords"
```
When one of those strings changes there is no unit test that detects that. I think we should have one const that will be used in both places.

**Pull request merged**